### PR TITLE
Feature: async rpc (#1706)

### LINKED ISSUE #1706: RPC Server: Make endpoints asynchronous
### Describe the feature
The non-administrative endpoints in the rpc server should be asynchronous. When a request is made, the rpc server should return immediately with an async token. Subsequent requests to the rpc server should be able to check the status of the running task via another endpoint, as well as retrieve task results when the request completes.

The following tasks should be **asynchronous**:
 - compile
 - run
 - compile_project
 - run_project
 - seed_project
 - test_project

The following tasks should be **synchronous**:
 - ps
 - kill
 - status
 - poll **new**

**Async tokens**
When an asynchronous request is made to the server, a `request_token` should be returned in the response. This should be something like a uuid.

The new `/poll` endpoint should accept a `request_token` either in the url (eg. `/poll/:request_token`) or as a query parameter (eg. `/poll?request_token=abc123`). My instinct is that this should be a POST request, but happy to discuss if something else makes more sense.

**The /poll endpoint**
The `/poll` endpoint should return a dictionary. The following examples show three different schemas for three different task states, but a single schema may be used if that is preferred in the implementation.

When the targeted request is running:
```
{
  "status": "running",
  "logs": [...]
}
```

When the targeted request has completed successfully:
```
{
  "status": "success",
  "data": {...},
  "logs": [...]
}
```

When the targeted request has completed with an error:
```
{
  "status": "error",
  "error": {...},
  "logs": [...]
}
```

**Storing task results**
The rpc server cannot store results indefinitely - it would _eventually_ run out of memory! Instead, the rpc server could serialize results & logs.... somewhere. When the `/poll` endpoint is queried for a completed task, the rpc server could load these results from that persistent storage location. See **Questions** below for a discussion....

**Questions:**
 - The implementation described here includes `logs` in the `/poll` response. Is this a reasonable way to fetch logs from a running task?
 - Can we re-purpose the `task_id` used internally (and exposed in the `ps`/`kill` commands) as the `request_token`? That feels sensible to me, but I'm not sure if there are any caveats to be aware of.
 - Is persisting results to disk (or somewhere else) viable and reasonable in the rpc server? Should we instead hold these results in memory, then purge them after some amount of time? Holding the results in memory is a path of lesser resistance, and we can implement async tasks separately from data persistence - the two questions are closely related, but definitely distinct and separable IMO. Very happy to discuss what our options are here.

@cmcarthur, @beckjake, would love to get your thoughts on these!

### Describe alternatives you've considered
we could ask users to keep their browsers open on a perfectly stable internet connection ;)

### PR DESCRIPTION
**Pull request by beckjake**
Fixes #1706 

Implement async tasks, mostly as described in #1706, with one big exception: On errors, an error object is returned directly instead of `"status": "error"` - it was just much easier to implement that way.

All the identified async methods now return a `request_token` instead of their results.
Adds a `poll` method that takes one mandatory argument, the `request_token`, and returns a `PollResult` on success, or the same error as before if the polled task got an error.
 - A `PollResult` always has a `status` field, which can be one of `['not started', 'initializing', 'running', 'success']`
 - if the status is `success`, it also comes with request-type-specific result data.
 - Polling for a missing request token is an error.
 

By default, garbage collection is checked on every json-rpc method:
  - `maxsize` defaults to 1000. If there are more than `maxsize` tasks, the oldest (`number_of_tasks` - `maxsize`) that have finished are cleaned up
 - `auto_reap_age` defaults to 30 days.
 - `reapsize` defaults to 500. If there are more than `reapsize` tasks, _all_ tasks that are older than the current `auto_reap_age` are garbage collected

All those numbers are totally invented by me and are very much arbitrary.

I also added a `gc` method that forces garbage collection and allows you to tweak GC settings:
  - the `task_ids` parameter, defaults to None and is a list of task IDs to garbage collect directly
  - the `before` parameter, defaults to None and is a datestamp. All tasks that finished after the datestamp will be garbage collected
  - the `settings` parameter, defaults to None and accepts an object with the below schema. This schema sets the internal GC tuning values.

GCSettings schema:
```
{
  'type': 'object'
  'additionalProperties': False,
  'properties': {
    'auto_reap_age': {'type': 'number'},
    'maxsize': {'type': 'integer'},
    'reapsize': {'type': 'integer'}
  },
  'required': ['maxsize', 'reapsize', 'auto_reap_age'],
}
```

### CODE REVIEW

File: core/dbt/config/profile.py [hunk 1]
```diff
@@ -20,7 +20,7 @@
 DEFAULT_THREADS = 1
 DEFAULT_PROFILES_DIR = os.path.join(os.path.expanduser('~'), '.dbt')
 PROFILES_DIR = os.path.expanduser(
-    os.environ.get('DBT_PROFILES_DIR', DEFAULT_PROFILES_DIR)
+    os.getenv('DBT_PROFILES_DIR', DEFAULT_PROFILES_DIR)
```
Comment (beckjake) [inline core/dbt/config/profile.py]: I changed all `os.environ.get` -> `os.getenv` for grep-ability reason when I was adding some secret magic threading-related flags to the RPC server.

File: core/dbt/contracts/results.py [hunk 2]
```diff
@@ -137,6 +147,15 @@ def write(self, path, omit_none=True):
...
         output.write(path, omit_none=omit_none)
 
+    def __len__(self):
+        return len(self.results)
+
+    def __iter__(self):
+        return iter(self.results)
+
+    def __getitem__(self, idx):
+        return self.results[idx]
```
Comment (beckjake) [inline core/dbt/contracts/results.py]: These changes make results a lot like the old thing where we returned `result.results`, which meant I didn't have to update a million tests. Also, it just seemed like a nice thing.

File: core/dbt/rpc/task_handler.py [hunk 3]
```diff
@@ -52,155 +110,236 @@ def _nt_setup(config, args):
     config.config.set_values(args.profiles_dir)
 
 
-def _task_bootstrap(task, queue, kwargs):
+def _task_bootstrap(
+    task: RPCTask,
+    queue,  # typing: Queue[Tuple[QueueMessageType, Any]]
```
[the remark this answers is not in the document]
Comment (beckjake) [reply core/dbt/rpc/task_handler.py]: This is a very annoying one. Basically, the built-in stdlib `queue.Queue` doesn't have any type-related information attached to it. So `typeshed`, which is included in `mypy`, provides its own `queue.Queue` for `mypy` to use that does support type information. It's all very annoying, as it means that `Queue[...]` isn't valid syntax so I can't use the `name: Type` syntax.

**Comment by beckjake** [comment]
> it looks like some types of request validation errors only show up in the poll method, but I would expect them to fail at request-time. One such example: submitting a run task without specifying a name. I'd expect the request to fail, but it's accepted and the failure is only apparent when we poll for the request ("handle_request() missing 1 required positional argument: 'name'"). Is this a minor change, or is there a good reason why it should work the way it's implemented here?

It's not a minor change really, as tasks define their arguments in `handle_request`, which is what the RPC server calls at runtime. To fix this, dbt rpc tasks would have to be implemented in two stages: one that takes the arguments and does some level of validation, and one that does the actual work.

**Comment by beckjake** [comment]
> the elapsed time in ps shows up as negative -- I see values like -14349.432432. Would that also impact garbage collection / task reaping?

Oops, this is a timezone bug! I think it would, yes.

File: core/dbt/rpc/task_handler.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (drewbanin) [reply core/dbt/rpc/task_handler.py]: roger! Thanks for the explanation

**Comment by beckjake** [comment]
>I'm seeing a 500 error when I hit the poll method for one particular project. I'm not sure exactly what's going wrong, but I can investigate further. This happens for run_project tasks, but not run tasks. I think this has something to do with control characters in the log lines. 

I can't reproduce this issue. I made a project that emits the same warning by adding a dummy dbt_project.yml line, and then did `dbt rpc`, started the task with `run_project`, and hit the `poll` endpoint repeatedly. Is there more to this.

**Comment by drewbanin** [comment]
> To fix this, dbt rpc tasks would have to be implemented in two stages: one that takes the arguments and does some level of validation, and one that does the actual work

Ok! Let's do that in this PR

**Pull request merged**